src/components/CallView/Grid/Grid.vue

```diff
@@ -828,6 +922,82 @@ export default {
 			return callParticipantModel.attributes.peerId === this.$store.getters.selectedVideoPeerId
 		},
 
+		isModelWithVideo(callParticipantModel) {
+			return callParticipantModel.attributes.videoAvailable
+				&& this.sharedDatas[callParticipantModel.attributes.peerId].remoteVideoBlocker.isVideoEnabled()
+				&& (typeof callParticipantModel.attributes.stream === 'object')
+		},
+
+		isModelWithAudio(callParticipantModel) {
+			const participant = this.$store.getters.getParticipantBySessionId(this.token, callParticipantModel.attributes.nextcloudSessionId)
+			if (!participant) {
+				return false
+			}
+			return participant?.permissions & PARTICIPANT.PERMISSIONS.PUBLISH_AUDIO
+		},
+
+		unpromoteSpeaker(model) {
+			const id = model.attributes.nextcloudSessionId
+			// remove the model from the temp promoted speakers
+			const index = this.videosTempPromoted.indexOf(model)
+			if (index !== -1) {
+				this.videosTempPromoted.splice(index, 1)
+			}
+
+			if (this.isModelWithVideo(model)) {
+				if (!this.tilesWithVideoMask.includes(id)) {
+					this.tilesWithVideoMask.unshift(id)
+				}
+			} else {
+				if (!this.tilesWithAudioMask.includes(id)) {
+					this.tilesWithAudioMask.unshift(id)
+				}
+			}
+		},
+
+		promoteSpeaker(model) {
+			const id = model.attributes.nextcloudSessionId
+
+			// if model is already in the first page, do nothing
+			if (this.orderedVideos.slice(0, this.slots).find(video => video.attributes.nextcloudSessionId === model.attributes.nextcloudSessionId)) {
+				return
+			}
+
+			if (this.modelsWithScreenshare.includes(model)) {
+				// tiles with screenshare have a better priority position already
+				// do nothing
+				return
+			}
+
+			// add the model
+			if (!this.videosTempPromoted.includes(model)) {
+				const videoIndex = this.tilesWithVideoMask.indexOf(id)
+				if (videoIndex !== -1) {
+					this.tilesWithVideoMask.splice(videoIndex, 1)
+				}
+				const audioIndex = this.tilesWithAudioMask.indexOf(id)
+				if (audioIndex !== -1) {
+					this.tilesWithAudioMask.splice(audioIndex, 1)
+				}
+
+				this.videosTempPromoted.unshift(model)
+			}
+		},
+
+		getOrderedTiles(tiles, orderMask) {
+			// Create a map for quick lookup
+			const tilesMap = new Map(tiles.map(tile => [tile.attributes.nextcloudSessionId, tile]))
```

Comment:
Since we iterate the models in computed one by one, we can provide a Map from there, and don't recalculate it again in method:
```suggestion
		getOrderedTiles(tilesMap, orderMask) {
```